proposals/0000-swift-settings.md

```diff
@@ -0,0 +1,446 @@
+arguments to a `#SwiftSettings` macro. Support for individual command line
+options is added to `SwiftSetting` by defining a static method on `SwiftSetting`
+in an extension. By using static methods and extensions in this manner, we are
+able to make `SwiftSetting` extensible without requiring the standard library to
+be updated whenever a new options is added since the extension can be defined in
+other modules. As an example, we add support for `defaultIsolation` via an
+extension in the Concurrency module as follows:
+
+```swift
+extension SwiftSetting {
+  public static func defaultIsolation(_ isolation: Actor.Type?) -> SwiftSetting { SwiftSetting() }
+}
+```
+
+Since this extension is in the `Concurrency` module, we are able to use
+`Actor.Type?` to specify the default isolation to be used which would not be
+possible if defaultIsolation was defined in the standard library since `Actor`
+is defined in Concurrency,
+
+Each static method on `SwiftSetting` is expecting to return a `SwiftSetting`
+struct. The actual value returned is not important since this code will not be
+executed.
+
+`#SwiftSetting` will only support an explicit opted-in group of command line
+flags. This initially will only include default isolation. All such options must
+possess the following characteristics to be compatible with `#SwiftSettings`:
+
+* No impact on parsing
+
+* Only modify the program in a manner that can be reflected in a textual interface
+  without the need for `#SwiftSettings` to be serialized into a textual
+  interface file.
+
+* Does not cause module wide effects that cannot be constrained to a single
+  file. An example of such a setting is `enable-library-evolution`.
+
+In the future, more options can be added as appropriate (see future
+directions). In order to ensure that any such options also obey the above
+restrictions, we expect that any feature that wants to be supported by
+`#SwiftSettings` must as part of their swift-evolution process consider the ABI
+and source stability impacts of adding support to `#SwiftSettings`. At minimum,
+the feature must obey the above requirements.
+
+Each argument can be passed to `#SwiftSettings` exactly once. Otherwise, an
+error will be emitted:
+
+```swift
+#SwiftSetting(
+  .defaultIsolation(MainActor.self),
+  .defaultIsolation(nil) // Error!
+)
+
+#SwiftSetting(
+  .defaultIsolation(MainActor.self) // Error!
+)
+```
+
+If a command line option is set to different values at the module and file
+level, the setting at the file level will take precedence.
+
+## Source compatibility
+
+The addition of the `#SwiftSettings` and `SwiftSetting` cannot impact the
+parsing of other constructs since the name shadowing rules in the compiler will
+ensure that any local macros or types that shadow those names will take
+precedence. If the user wishes to still use `#SwiftSettings`, the user can spell
+the macro as `#Swift.SwiftSettings` as needed.
```

Comment:
Does this work because it's parsed as a freestanding declaration macro? I'm pretty sure that in expression context, `#Swift.SwiftSettings` is parsed as a `#Swift` freestanding expression macro with `.SwiftSettings` as a postfix member. (doesn't really matter for this proposal, only for my curiosity!)